core/server/common/src/main/java/alluxio/master/journal/raft/SnapshotReplicationManager.java

```diff
@@ -449,75 +479,59 @@ private void requestSnapshotFromFollowers() {
-        }
-        LOG.debug("Received snapshot info from follower {} - {}", peerId, response);
-        if (!response.hasSnapshotInfoResponse()) {
-          LOG.warn("Invalid response for GetSnapshotInfoRequest {}", response);
-          continue;
-        }
-        SnapshotMetadata latest = response.getSnapshotInfoResponse().getLatest();
-        if (latest == null) {
-          LOG.debug("Follower {} does not have a snapshot", peerId);
-          continue;
-        }
-        if (snapshotMetadata == null
-            || (latest.getSnapshotTerm() >= snapshotMetadata.getSnapshotTerm())
-            && latest.getSnapshotIndex() > snapshotMetadata.getSnapshotIndex()) {
-          snapshotMetadata = latest;
-          snapshotOwner = peerId;
+          LOG.debug("Received snapshot info from follower {} - {}", peerId, response);
+          if (!response.hasSnapshotInfoResponse()) {
+            throw new IOException("Invalid response for GetSnapshotInfoRequest " + response);
+          }
+          SnapshotMetadata latest = response.getSnapshotInfoResponse().getLatest();
+          if (snapshotMetadata == null
+              || (latest.getSnapshotTerm() >= snapshotMetadata.getSnapshotTerm())
+              && latest.getSnapshotIndex() > snapshotMetadata.getSnapshotIndex()) {
+            mSnapshotCandidates.add(new Pair<>(latest, peerId));
+          }
+        } catch (Exception e) {
+          LOG.warn("Error while requesting snapshot info from {}: {}", peerId, e.toString());
         }
       }
-      if (snapshotOwner == null) {
-        throw new UnavailableException("No recent snapshot found from followers");
-      }
     } catch (Exception e) {
       LogUtils.warnWithException(LOG, "Failed to request snapshot info from followers", e);
-      transitionState(DownloadState.REQUEST_INFO, DownloadState.IDLE);
-      return;
-    }
-    // we have a follower with a more recent snapshot, request an upload
-    LOG.info("Request snapshot data from follower {}", snapshotOwner);
-    mSnapshotRequestTime = CommonUtils.getCurrentMs();
-    transitionState(DownloadState.REQUEST_INFO, DownloadState.REQUEST_DATA);
-    try {
-      RaftClientReply reply = mJournalSystem.sendMessageAsync(snapshotOwner,
-          toMessage(JournalQueryRequest.newBuilder()
-              .setSnapshotRequest(GetSnapshotRequest.getDefaultInstance()).build()))
-          .get();
-      if (reply.getException() != null) {
-        throw reply.getException();
-      }
-    } catch (Exception e) {
-      LOG.error("Failed to request snapshot data from {}", snapshotOwner, e);
-      transitionState(DownloadState.REQUEST_DATA, DownloadState.IDLE);
     }
   }
 
-  private void checkRequestTimeout() {
-    if (CommonUtils.getCurrentMs() - mSnapshotRequestTime > SNAPSHOT_REQUEST_TIMEOUT_MS) {
-      transitionState(DownloadState.REQUEST_DATA, DownloadState.IDLE);
+  private boolean requestData() {
+    Preconditions.checkState(mDownloadState.get() == DownloadState.REQUEST_DATA);
```

Comment:
if an upload fails, what happens next? since we don't reset state to IDLE, I assume the requestData() is called again. Then wouldn't it request the snapshot from the same follower?

Reply:
If an upload fails, the `SnapshotReplicationManager` moves into `REQUEST_DATA` state and calls `requestSnapshotFromFollowers` ([here](https://github.com/Alluxio/alluxio/pull/14934/files#diff-835220c648a2185d56f3d9676b7077bd280bfa13e7f305a3093aa5d7c87028beR302)). 
This calls `requestData()` right away and requests snapshots from other followers it had polled information from before. If no other followers are able to upload, it moves back to `IDLE` where it will subsequently request new information about snapshots from followers.

Reply:
Oh I see, you're polling from the queue. Thanks for the explanation.